src/main/java/de/tum/in/www1/artemis/service/scheduled/TextClusteringScheduleService.java

```diff
@@ -0,0 +1,85 @@
+package de.tum.in.www1.artemis.service.scheduled;
+
+import java.time.ZonedDateTime;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ScheduledFuture;
+
+import javax.annotation.PostConstruct;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import de.tum.in.www1.artemis.domain.TextExercise;
+import de.tum.in.www1.artemis.domain.enumeration.ExerciseLifecycle;
+import de.tum.in.www1.artemis.security.SecurityUtils;
+import de.tum.in.www1.artemis.service.ExerciseLifecycleService;
+import de.tum.in.www1.artemis.service.TextClusteringService;
+import de.tum.in.www1.artemis.service.TextExerciseService;
+
+@Service
+@Profile("automaticText")
+public class TextClusteringScheduleService {
+
+    private final Logger log = LoggerFactory.getLogger(TextClusteringScheduleService.class);
+
+    private final ExerciseLifecycleService exerciseLifecycleService;
+
+    private final TextExerciseService textExerciseService;
+
+    private final Map<Long, ScheduledFuture> scheduledClusteringTasks = new HashMap<>();
+
+    private final TextClusteringService textClusteringService;
+
+    public TextClusteringScheduleService(ExerciseLifecycleService exerciseLifecycleService, TextExerciseService textExerciseService, TextClusteringService textClusteringService) {
+        this.exerciseLifecycleService = exerciseLifecycleService;
+        this.textExerciseService = textExerciseService;
+        this.textClusteringService = textClusteringService;
+    }
+
+    @PostConstruct
+    private void scheduleRunningExercisesOnStartup() {
+        List<TextExercise> runningTextExercises = textExerciseService.findAllAutomaticAssessmentTextExercisesWithFutureDueDate();
+        runningTextExercises.forEach(this::scheduleExerciseForClustering);
+        log.info("Scheduled text clustering for " + runningTextExercises.size() + " text exercises with future due dates.");
+    }
+
+    public void scheduleExerciseForClusteringIfRequired(TextExercise exercise) {
+        if (!exercise.isAutomaticAssessmentEnabled()) {
+            cancelScheduledClustering(exercise);
```

Comment:
Why does the Clustering has to be canceled if it isn't even started?

Reply:
A check within the `cancelScheduledClustering` method. The idea is, that you can only ever schedule clustering once for every exercise. I will extend the Java Doc.